### ISSUE TITLE
No way to access the literal type at a specific object key?

### ISSUE THREAD

**Issue by callumlocke** [facebook/flow#6651]
I'm trying to access the literal type of the value at a specific key of a given object, in order to use with generics to describe the complex relationship between a function's input and return type.

If you freeze an object, you can use the `$Values` utility to get a union of [literal types](https://flow.org/en/docs/types/literals/) based on an object's values:

```jsx
const foo = Object.freeze({
  a: 'bar',
  b: 123,
  c: true,
})

type MyType = $Values<typeof foo> // 'bar' | 123 | true

('bar': MyType)  // ok
(123: MyType)    // ok
('nope': MyType) // flow error
```

That works fine. But what if I want to get only one literal type, at a specific key, `foo.a` (i.e. `'bar'` in this case)? I've tried several ways but they all lose accuracy, becoming the primitive `string` type:

```jsx
type Attempt1 = typeof foo.a // string

type Attempt2 = $PropertyType<typeof foo, 'a'> // string

type Attempt3 = $ElementType<typeof foo, 'a'> // string

type Attempt4 = $Values<
  $Diff<
    typeof foo,
    $Diff<typeof foo, { a: string }>
  >
> // string
```

Is there any way to do it?

**Comment by fishythefish** [comment]
You can annotate the type of `foo` or make the type of `foo.a` explicit: `a: ('bar': 'bar')`.

**Comment by callumlocke** [comment]
@fishythefish thanks but I need a way to do it dynamically for an object that is not known in advance, accessed only via via a [generic](https://flow.org/en/docs/types/generics/) reference.

**Comment by bradennapier** [comment]
Yeah this would appear to be a bug.  Most likely when they allowed the unions via Object.freeze they did not extenf the interface to handling other type calls like $PropertyType, etc. 

In this case, Flow should absolutely know it will be 'bar' and never anything else because it is no longer possible for foo.a to be any other value than 'bar'. 

These are the bugs that really hold Flow back and make Typescript more appealing for sure.

**Comment by gkz** [comment]
Original example now works as you expect, as behavior of `Object.freeze` was updated: https://flow.org/try/#1N4Igxg9gdgZglgcxALlAIwIZoKYBsD6uEEAztvhgE6UYCe+JADpdhgCYowa5kA0I2KAFcAtiRQAXSkOz9sADwxgJ+NPTbYuQ3BMnTZA+Y2yU4IwRO4A6SFBIrGVDGM7c+h46fNRLuKxJIGWh8MeT0ZfhYlCStpHzNsFBAMIQkIEQwJODAQfiEyfBE4eWw2fDgofDBMsAALfAA3KjgsXGxxZC4eAw0G-GhcWn9aY3wWZldu-g1mbGqJUoBaCRHEzrcDEgBrbAk62kXhXFxJ923d-cPRHEpTgyEoMDaqZdW7vKgoOfaSKgOKpqmDA+d4gB5fMA-P6LCCMLLQbiLOoYCqgh6-GDYRYIXYLSgkRZkCR4jpddwPfJLZjpOBkO4AX34kA0SVs9gABDBiOyALzsgDyaAAVnMYjAWNgAF7YAAUwAAOlB2eyMMh2QByTCUdW8RXKtBqgCMACYAMy6pXssBqqQRRX0gCUisVK2M7IAsrQACqrXnsgAkADVuDISAAeV3YCAwTnEAB87IA9ImNVr1eyAD7sk2mzPs23YZ1QGWaqjpjAkD3e1YOgDcyuT7IgW0VMpzKsrnp9xjryobKebrfVUFh2HLnerPfrjZgRAA7uyTJQIJQi5H2QBBYnYERww1+yPR2MQKwYacp0vate+rcLXcSY1+-0ABWXnhW3ewEdWR65EF4GoYOqCaNpe6rXm6t47nCuZ8v6ACibTeBIn7fsYv7EAB6pASBF5phB2Cbtu94ACxPsGuChmGeoBgAInAMAwNRlrKoeMZ-hafZ0QxTFsceAHACqar2KYUAIOy9JxjRUlQLhqZlrkIANCYJBwNASQNAADFYxoAGymlYmkgPSQA
